crates/schema-api/src/lib.rs

```diff
@@ -94,7 +100,128 @@ pub mod deployment {
+                serde(
+                    default,
+                    with = "serde_with::As::<Option<restate_serde_util::VersionSerde>>"
+                )
+            )]
+            // this field did not used to be stored, so we must consider it optional when deserialising
+            http_version: Option<http::Version>,
+        },
+        Lambda {
+            arn: LambdaARN,
+            assume_role_arn: Option<ByteString>,
+        },
+    }
+
+    #[cfg(feature = "serde")]
+    impl From<DeploymentTypeShadow> for DeploymentType {
+        fn from(value: DeploymentTypeShadow) -> Self {
+            match value {
+                DeploymentTypeShadow::Http {
+                    address,
+                    protocol_type,
+                    http_version,
+                } => Self::Http {
+                    address,
+                    protocol_type,
+                    http_version: match http_version {
+                        Some(v) => v,
+                        None => Self::backfill_http_version(protocol_type),
+                    },
+                },
+                DeploymentTypeShadow::Lambda {
+                    arn,
+                    assume_role_arn,
+                } => Self::Lambda {
+                    arn,
+                    assume_role_arn,
+                },
+            }
+        }
+    }
+
+    #[cfg(test)]
+    #[cfg(feature = "serde")]
+    mod serde_tests {
+        use bytestring::ByteString;
+        use http::Uri;
+        use restate_types::{identifiers::LambdaARN, storage::StorageCodec};
+
+        use crate::deployment::{DeploymentType, ProtocolType};
+
+        #[derive(serde::Serialize, serde::Deserialize)]
+        enum OldDeploymentType {
+            Http {
+                #[serde(with = "serde_with::As::<serde_with::DisplayFromStr>")]
+                address: Uri,
+                protocol_type: ProtocolType,
+            },
+            Lambda {
+                arn: LambdaARN,
+                assume_role_arn: Option<ByteString>,
+            },
+        }
+
+        restate_types::flexbuffers_storage_encode_decode!(OldDeploymentType);
+        restate_types::flexbuffers_storage_encode_decode!(DeploymentType);
+
+        #[test]
+        fn can_deserialise_without_http_version() {
+            let mut buf = bytes::BytesMut::default();
+            StorageCodec::encode(
+                &OldDeploymentType::Http {
+                    address: Uri::from_static("google.com"),
+                    protocol_type: ProtocolType::BidiStream,
+                },
+                &mut buf,
+            )
+            .unwrap();
+            let dt: DeploymentType = StorageCodec::decode(&mut buf).unwrap();
+            let serialised = serde_json::to_string(&dt).unwrap();
+            assert_eq!(
+                r#"{"Http":{"address":"google.com","protocol_type":"BidiStream","http_version":"HTTP/2.0"}}"#,
+                serialised
+            );
```

Comment:
I guess you could also let `DeploymentType` derive `PartialEq` and then `assert_eq!(dt, DeploymentType { ... })` w/o having to go through json and a string comparison.